"""РАБОТА С ОШИБКАМИ"""

import sys
from logging import exception

# представим, что у нас не локальные док-ты, а файлы по url адресам:
url_list = ["text1.txt", "text2.txt", "text3.txt", "text4.txt"]

# создадим сразу 2 списка, которые будем заполнять данными,
# в одном списке - содержимое файлов, во втором список файлов, которые не смогли прочесть,
# из-за какой-то ошибки при чтении
list_defect = []
list_info = []

try:
    # для начала пройдем циклом for, по списку из url адресов
    for url in url_list:
        # нам нужно открыть файлы по url адресам, что перебирает цикл
        # создадим переменную, для открытия файлов
        # r = open(url, encoding="utf-8")
        # print(r.read())
        # получаем ошибку, что бы ее исключить, запишем код через оператор try
        try:
            r = open(url, encoding="utf-8")
            # сразу будем читать файлы, добавляя их содержимое в список list_info
            list_info.append(r.read())
            # print(list_info)
        # на случай ошибки, запишем оператор
        except Exception:
            # если сработает этот оператор, то будем добавлять в список list_defect, название файла с ошибкой
            list_defect.append(url)
            print(f"{list_defect}, этот файл не читается")
            print(list_info)
            # на случай, непредвиденной ошибки, используем оператор exit из метода sys
            sys.exit()
            # как только мы закрыли программу sys.exit(), все данные в памяти были стерты и потеряны,
            # что бы ничего не потерять, весь этот цикл, помещаем в еще один try строка 15,

        continue
    # и запишем оператор finally: и запишем в нем код, который сохранит все наши данные,
    # (finally: срабатывает всегда)
finally:
    # что бы сохранить все данные из кода, создадим текстовый документ
    r = open("save.txt", "a", encoding="utf-8")
    # пройдемся по списку с инфой
    for i in list_info:
        r.write(i)
    r.write(str(list_defect))
    r.close()
    print("данные не потеряны")
